qa/rpc-tests/zcjoinsplit.py

```diff
@@ -4,53 +4,112 @@
 
 class JoinSplitTest(BitcoinTestFramework):
     def setup_network(self):
         self.nodes = []
         self.is_network_split = False
         self.nodes.append(start_node(0, self.options.tmpdir))
+        self.nodes.append(start_node(1, self.options.tmpdir))
+        connect_nodes(self.nodes[0], 1)
 
     def run_test(self):
-        zckeypair = self.nodes[0].zcrawkeygen()
-        zcsecretkey = zckeypair["zcsecretkey"]
-        zcaddress = zckeypair["zcaddress"]
 
-        (total_in, inputs) = gather_inputs(self.nodes[0], 45.75)
-        protect_tx = self.nodes[0].createrawtransaction(inputs, {})
-        joinsplit_result = self.nodes[0].zcrawjoinsplit(protect_tx, {}, {zcaddress:45.74}, 45.74, 0)
+        ForkHeight = ForkHeights['SHIELDED_POOL_DEPRECATION']
+
+        # first round pre-fork, second round post-fork
+        pre_fork_round = 0
+        just_pre_fork_round = 1
+        post_fork_round = 2
+        for round in range(3):
+            zckeypair = self.nodes[0].zcrawkeygen()
+            zcsecretkey = zckeypair["zcsecretkey"]
+            zcaddress = zckeypair["zcaddress"]
+
+            (total_in, inputs) = gather_inputs(self.nodes[0], 45.75)
+
+            protect_tx = self.nodes[0].createrawtransaction(inputs, {})
+            joinsplit_result = self.nodes[0].zcrawjoinsplit(protect_tx, {}, {zcaddress: total_in - Decimal(0.01)}, total_in - Decimal(0.01), 0)
+
+            receive_result = self.nodes[0].zcrawreceive(zcsecretkey, joinsplit_result["encryptednote1"])
+            assert_equal(receive_result["exists"], False)
+
+            protect_tx = self.nodes[0].signrawtransaction(joinsplit_result["rawtxn"])
+
+            # disconnect the nodes so that the chains are not synced
+            if (round == just_pre_fork_round):
+                disconnect_nodes(self.nodes[0], 1)
+
+            try:
+                self.nodes[0].sendrawtransaction(protect_tx["hex"])
+                if (round == post_fork_round):
+                    assert(False)
+            except JSONRPCException as e:
+                if (round == pre_fork_round or round == just_pre_fork_round):
+                    print("Unexpected exception caught during testing: " + str(sys.exc_info()[0]))
+                    assert(False)
+                else:
+                    print("Expected exception caught during testing due to deprecation (error=" + str(e.error["code"]) + ")")
+                    assert_equal(e.error["code"], RPC_VERIFY_REJECTED)
+                    continue # other steps of this round are skipped
+
+            if (round == just_pre_fork_round):
+                # advance chain on node1 (so that shielded pool deprecation hard fork is active)
+                self.nodes[1].generate(20)
+                connect_nodes(self.nodes[0], 1)
+                sync_blocks(self.nodes)
+                # make sure now deprecated tx is in node0 mempool but not in node1 mempool
+                assert_equal(self.nodes[0].getmempoolinfo()["size"], 1)
```

Comment:
Currently, if a transaction already contained in the mempool becomes irrevocably invalid in terms of contextual check, it would remain indefinitely in the mempool, the only way of getting rid of it being a node restart. I have checked, also for the hard fork PHGR->GROTH no feature involving mempool cleaning was provided. This improvement activity is tracked and will be performed in future sprint.